<a href="https://colab.research.google.com/github/torchflash/MMAI831-AIOS-Assignment/blob/main/Irene_AIOS_KNN_%26_Parametric_Model.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Objective: To build parametric and machine learning models for predicting sales volume using various independent variables, and to discuss the advantages and disadvantages of each model in terms of inference, predictability, and the estimation process.
Dataset: The dataset "MMM_store sales.xlsx" contains information about sales volume and the independent variables of interest, including In Store Advertising ($), Billboard Advertising ($), Print Advertising ($), and control variables such as Price ($), Satisfaction, and Competitor Advertising ($).
Details:
	You can do the analysis in any software of your choice.
	Use tables, figures and other visuals to increase the readability and effectiveness of your report. 
	Submit only one assignment for your team via Dropbox. Any team member can make the submission. Make sure to have all team members' names on the report.


In [1]:
import pandas as pd

# Load the dataset
data = pd.read_excel("/content/MMM_Store sales.xlsx")

# Display the first few rows of the dataset
print(data.head())


   Month  In store advertising ($)  billboard Advertising ($)  \
0      1                      1698                        698   
1      2                      2069                       1075   
2      3                      2271                        615   
3      4                      1413                       1158   
4      5                      2107                       1175   

   Print Advertising ($)  Satisfaction  Competitor Adversiting ($)  Price ($)  \
0                    556            68                         607         95   
1                    775            72                         637        101   
2                    772            71                         761         95   
3                   1098            79                        1183         98   
4                    386            77                         940         97   

      Sales ($)  
0  13605.611909  
1  19142.673338  
2  15027.623400  
3  16589.659102  
4  21887.582100  


In [2]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score, mean_squared_error

# Load the dataset
data = pd.read_excel("/content/MMM_Store sales.xlsx")

# Split the dataset into training and testing sets
train_size = int(0.75 * len(data))
sales_train = data[:train_size]
sales_test = data[train_size:]

# Split the training set into independent variables (X_train) and dependent variable (y_train)
X_train = sales_train[['In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Price ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y_train = sales_train['Sales ($)']

# Split the testing set into independent variables (X_test) and dependent variable (y_test)
X_test = sales_test[['In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Price ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y_test = sales_test['Sales ($)']




In [3]:
# Split the dataset into training and testing sets
train_size = int(0.75 * len(data))
sales_train = data[:train_size]
sales_test = data[train_size:]

# Split the training set into independent variables (X_train) and dependent variable (y_train)
X_train = sales_train[['In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Price ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y_train = sales_train['Sales ($)']

# Split the testing set into independent variables (X_test) and dependent variable (y_test)
X_test = sales_test[['In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Price ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y_test = sales_test['Sales ($)']




In [4]:
# Parametric Model: Linear Regression
parametric_model = LinearRegression()
parametric_model.fit(X_train, y_train)

# Machine Learning Model: k-Nearest Neighbors Regression
ml_model = KNeighborsRegressor()
ml_model.fit(X_train, y_train)

# Predict sales volume using the parametric model
parametric_predictions = parametric_model.predict(X_test)

# Predict sales volume using the machine learning model
ml_predictions = ml_model.predict(X_test)




In [5]:
# Calculate evaluation metrics for the parametric model
parametric_r2 = r2_score(y_test, parametric_predictions)
parametric_mse = mean_squared_error(y_test, parametric_predictions)
parametric_mapd = np.mean(np.abs((y_test - parametric_predictions) / y_test)) * 100

# Calculate evaluation metrics for the machine learning model
ml_r2 = r2_score(y_test, ml_predictions)
ml_mse = mean_squared_error(y_test, ml_predictions)
ml_mapd = np.mean(np.abs((y_test - ml_predictions) / y_test)) * 100

# Print the evaluation metrics for both models
print("Parametric Model Evaluation:")
print("R-squared (R2):", parametric_r2)
print("Mean Squared Error (MSE):", parametric_mse)
print("Mean Absolute Percentage Difference (MAPD):", parametric_mapd)

print("\nMachine Learning Model (KNN) Evaluation:")
print("R-squared (R2):", ml_r2)
print("Mean Squared Error (MSE):", ml_mse)
print("Mean Absolute Percentage Difference (MAPD):", ml_mapd)


Parametric Model Evaluation:
R-squared (R2): 0.9201816438077914
Mean Squared Error (MSE): 1113983.9557694958
Mean Absolute Percentage Difference (MAPD): 4.9443513489899065

Machine Learning Model (KNN) Evaluation:
R-squared (R2): 0.7245906707595942
Mean Squared Error (MSE): 3843747.0862498065
Mean Absolute Percentage Difference (MAPD): 9.231007537324638


In [6]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LinearRegression
from sklearn.neighbors import KNeighborsRegressor
from sklearn.metrics import r2_score, mean_squared_error
import statsmodels.api as sm

# Load the dataset
data = pd.read_excel("/content/MMM_Store sales.xlsx")

# Split the dataset into training and testing sets
train_size = int(0.75 * len(data))
sales_train = data[:train_size]
sales_test = data[train_size:]


In [7]:
# Add pairwise interactions for advertising channels
sales_train['Store_Billboard'] = sales_train['In store advertising ($)'] * sales_train['billboard Advertising ($)']
sales_train['Store_Printout'] = sales_train['In store advertising ($)'] * sales_train['Print Advertising ($)']
sales_train['Billboard_Printout'] = sales_train['billboard Advertising ($)'] * sales_train['Print Advertising ($)']

sales_test['Store_Billboard'] = sales_test['In store advertising ($)'] * sales_test['billboard Advertising ($)']
sales_test['Store_Printout'] = sales_test['In store advertising ($)'] * sales_test['Print Advertising ($)']
sales_test['Billboard_Printout'] = sales_test['billboard Advertising ($)'] * sales_test['Print Advertising ($)']


<ipython-input-7-af173bce6ddb>:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sales_train['Store_Billboard'] = sales_train['In store advertising ($)'] * sales_train['billboard Advertising ($)']
<ipython-input-7-af173bce6ddb>:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  sales_train['Store_Printout'] = sales_train['In store advertising ($)'] * sales_train['Print Advertising ($)']
<ipython-input-7-af173bce6ddb>:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.

In [8]:
# Training set
X_train = sales_train[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction',
                       'Competitor Adversiting ($)', 'Store_Billboard', 'Store_Printout', 'Billboard_Printout']]
y_train = sales_train['Sales ($)']

# Testing set
X_test = sales_test[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction',
                     'Competitor Adversiting ($)', 'Store_Billboard', 'Store_Printout', 'Billboard_Printout']]
y_test = sales_test['Sales ($)']


In [9]:
# Parametric Model: Linear Regression with interactions
parametric_model = LinearRegression()
parametric_model.fit(X_train, y_train)

# Perform inference using the parametric model
X_train_inference = sm.add_constant(X_train)  # Add a constant term for intercept
parametric_results = sm.OLS(y_train, X_train_inference).fit()
parametric_coefficients = parametric_results.params
parametric_standard_errors = parametric_results.bse


In [10]:
# Machine Learning Model: k-Nearest Neighbors Regression
ml_model = KNeighborsRegressor()
ml_model.fit(X_train, y_train)


KNeighborsRegressor()

In [11]:
# Predict sales volume using the parametric model
parametric_predictions = parametric_model.predict(X_test)

# Predict sales volume using the machine learning model
ml_predictions = ml_model.predict(X_test)


In [12]:
# Calculate evaluation metrics for the parametric model
parametric_r2 = r2_score(y_test, parametric_predictions)
parametric_mse = mean_squared_error(y_test, parametric_predictions)
parametric_mapd = np.mean(np.abs((y_test - parametric_predictions) / y_test)) * 100



In [13]:
# Calculate evaluation metrics for the machine learning model
ml_r2 = r2_score(y_test, ml_predictions)
ml_mse = mean_squared_error(y_test, ml_predictions)
ml_mapd = np.mean(np.abs((y_test - ml_predictions) / y_test)) * 100

# Print the evaluation metrics
print("Parametric Model:")
print("R-squared:", parametric_r2)
print("MSE:", parametric_mse)
print("MAPD:", parametric_mapd)

print("\nMachine Learning Model (KNN):")
print("R-squared:", ml_r2)
print("MSE:", ml_mse)
print("MAPD:", ml_mapd)

Parametric Model:
R-squared: 0.9283838439918626
MSE: 999510.0447174818
MAPD: 4.803844114352636

Machine Learning Model (KNN):
R-squared: 0.7502883567067267
MSE: 3485097.632525485
MAPD: 8.546523617700116


In [14]:
from sklearn.neighbors import KNeighborsRegressor

# Define the independent variables and target variable
X = data[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y = data['Sales ($)']
# Initialize and fit the KNN model
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X, y)

# Make predictions on the same dataset
y_pred = knn.predict(X)

# Calculate ROI
cost = data['In store advertising ($)'] + data['billboard Advertising ($)'] + data['Print Advertising ($)']
roi = (y - cost) / cost

# Calculate Elasticity
percentage_change_sales = ((y - y_pred) / y_pred) * 100
percentage_change_price = ((data['Price ($)'] - data['Price ($)'].shift(1)) / data['Price ($)'].shift(1)) * 100
elasticity = percentage_change_sales / percentage_change_price

# Add ROI and Elasticity to the dataset
data['ROI'] = roi
data['Elasticity'] = elasticity

# Print the updated dataset
print(data)


     Month  In store advertising ($)  billboard Advertising ($)  \
0        1                      1698                        698   
1        2                      2069                       1075   
2        3                      2271                        615   
3        4                      1413                       1158   
4        5                      2107                       1175   
..     ...                       ...                        ...   
995    996                      1945                       1065   
996    997                      2137                       1333   
997    998                      2120                       1061   
998    999                      2190                        656   
999   1000                      1888                        702   

     Print Advertising ($)  Satisfaction  Competitor Adversiting ($)  \
0                      556            68                         607   
1                      775            72           

In [15]:
from sklearn.neighbors import KNeighborsRegressor

# Define the independent variables and target variable
X = data[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
y = data['Sales ($)']

# Initialize and fit the KNN model
knn = KNeighborsRegressor(n_neighbors=5)
knn.fit(X, y)

# Make predictions on the same dataset
y_pred = knn.predict(X)

# Calculate ROI and Elasticity for each month
roi_list = []
elasticity_list = []

for i in range(len(data)):
    sales = data.loc[i, 'Sales ($)']
    cost = data.loc[i, 'In store advertising ($)'] + data.loc[i, 'billboard Advertising ($)'] + data.loc[i, 'Print Advertising ($)']
    
    roi = (sales - cost) / cost
    roi_list.append(roi)
    
    if i > 0:
        percentage_change_sales = ((sales - y_pred[i-1]) / y_pred[i-1]) * 100
        percentage_change_price = ((data.loc[i, 'Price ($)'] - data.loc[i-1, 'Price ($)']) / data.loc[i-1, 'Price ($)']) * 100
        
        elasticity = percentage_change_sales / percentage_change_price
        elasticity_list.append(elasticity)
    else:
        elasticity_list.append(None)

# Add ROI and Elasticity to the dataset
data['ROI'] = roi_list
data['Elasticity'] = elasticity_list
# Export the DataFrame to an Excel file
data.to_excel('output_data.xlsx', index=False)

# Print the updated dataset
print(data)


<ipython-input-15-0b6950907cf1>:29: RuntimeWarning: divide by zero encountered in double_scalars
  elasticity = percentage_change_sales / percentage_change_price


     Month  In store advertising ($)  billboard Advertising ($)  \
0        1                      1698                        698   
1        2                      2069                       1075   
2        3                      2271                        615   
3        4                      1413                       1158   
4        5                      2107                       1175   
..     ...                       ...                        ...   
995    996                      1945                       1065   
996    997                      2137                       1333   
997    998                      2120                       1061   
998    999                      2190                        656   
999   1000                      1888                        702   

     Print Advertising ($)  Satisfaction  Competitor Adversiting ($)  \
0                      556            68                         607   
1                      775            72           

In [16]:
import pandas as pd
from sklearn.linear_model import LinearRegression

# Load the Excel file
data = pd.read_excel('/content/output_data Elasticity python.xlsx')


# Drop rows with missing or invalid values
data = data.replace([np.inf, -np.inf], np.nan)
data = data.dropna(subset=['ROI', 'Elasticity'], how='any')

# Extract the independent variables and target variables
X = data[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
roi = data['ROI']
elasticity = data['Elasticity']

# Perform linear regression for ROI
regressor_roi = LinearRegression()
regressor_roi.fit(X, roi)
coefficients_roi = regressor_roi.coef_

# Perform linear regression for Elasticity
regressor_elasticity = LinearRegression()
regressor_elasticity.fit(X, elasticity)
coefficients_elasticity = regressor_elasticity.coef_

# Print the coefficients for ROI
print("Coefficients for ROI:")
for variable, coefficient in zip(X.columns, coefficients_roi):
    print(f"{variable}: {coefficient}")

# Print the coefficients for Elasticity
print("Coefficients for Elasticity:")
for variable, coefficient in zip(X.columns, coefficients_elasticity):
    print(f"{variable}: {coefficient}")







Coefficients for ROI:
Price ($): -0.05239205188935794
In store advertising ($): 0.0005833622628190582
billboard Advertising ($): 0.001648267463837007
Print Advertising ($): -0.0012171059606555683
Satisfaction: 0.05235932148768951
Competitor Adversiting ($): -0.0004109316547236338
Coefficients for Elasticity:
Price ($): 0.16457757495951436
In store advertising ($): -0.002481087171662698
billboard Advertising ($): -0.0019566119997896196
Print Advertising ($): -0.000800615465056024
Satisfaction: 0.017095954523260137
Competitor Adversiting ($): -0.0013383433408159038


For ROI:
Satisfaction has the greatest impact : An increase in satisfaction by 1 unit is associated with an increase in ROI by approximately 0.0524 units.


For Elasticity using data produced by code:

Price has the greatest impact ($): A 1% increase in price is associated with an increase in elasticity by approximately 0.1646 units.






Using Excel calcuated ROI and Elasticity
ROI most impactful feature :  Satisfaction
Elasticity most impactful feature: Price

In [17]:
import pandas as pd
from sklearn.linear_model import LinearRegression

# Load the Excel file
data = pd.read_excel('/content/output_data Elasticity calculated.xlsx')


# Drop rows with missing or invalid values
data = data.replace([np.inf, -np.inf], np.nan)
data = data.dropna(subset=['ROI', 'Elasticity'], how='any')

# Extract the independent variables and target variables
X = data[['Price ($)', 'In store advertising ($)', 'billboard Advertising ($)', 'Print Advertising ($)', 'Satisfaction', 'Competitor Adversiting ($)']]
roi = data['ROI']
elasticity = data['Elasticity']

# Perform linear regression for ROI
regressor_roi = LinearRegression()
regressor_roi.fit(X, roi)
coefficients_roi = regressor_roi.coef_

# Perform linear regression for Elasticity
regressor_elasticity = LinearRegression()
regressor_elasticity.fit(X, elasticity)
coefficients_elasticity = regressor_elasticity.coef_

# Print the coefficients for ROI
print("Coefficients for ROI:")
for variable, coefficient in zip(X.columns, coefficients_roi):
    print(f"{variable}: {coefficient}")

# Print the coefficients for Elasticity
print("Coefficients for Elasticity:")
for variable, coefficient in zip(X.columns, coefficients_elasticity):
    print(f"{variable}: {coefficient}")

Coefficients for ROI:
Price ($): -0.05239205188935794
In store advertising ($): 0.0005833622628190582
billboard Advertising ($): 0.001648267463837007
Print Advertising ($): -0.0012171059606555683
Satisfaction: 0.05235932148768951
Competitor Adversiting ($): -0.0004109316547236338
Coefficients for Elasticity:
Price ($): 0.24432693744673994
In store advertising ($): -0.006215858569084292
billboard Advertising ($): -0.0019453924617828663
Print Advertising ($): -0.0016715495025396002
Satisfaction: 0.017997729070869155
Competitor Adversiting ($): -0.002068115254107891
